flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,242 @@
+							if (cancelThrowable != null) {
+								log.warn("Cancellation of job {} failed", jobId, cancelThrowable);
+							}
+						});
+
+						// cancellation will eventually complete the jobResultFuture
+						jobResultFuture.whenComplete((archivedExecutionGraph, resultThrowable) -> {
+							synchronized (lock) {
+								if (resultThrowable == null) {
+									jobStatus = archivedExecutionGraph.getState();
+								} else {
+									jobStatus = JobStatus.FAILED;
+								}
+							}
+						});
+					} else {
+						jobStatus = JobStatus.RUNNING; // this status should never be exposed from the DispatcherJob. Only used internally for tracking running state
+					}
+				} else { // failure during initialization
+					if (submissionType == SubmissionType.RECOVERY) {
+						jobResultFuture.completeExceptionally(throwable);
+					} else {
+						jobResultFuture.complete(ArchivedExecutionGraph.createFromInitializingJob(
+							jobId,
+							jobName,
+							throwable,
+							JobStatus.FAILED,
+							initializationTimestamp));
+					}
+				}
+			}
+			return null;
+		}));
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> getResultFuture() {
+		return jobResultFuture;
+	}
+
+	public CompletableFuture<JobDetails> requestJobDetails(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobDetails(
+					timeout));
+			} else {
+				int[] tasksPerState = new int[ExecutionState.values().length];
+				return CompletableFuture.completedFuture(new JobDetails(
+					jobId,
+					jobName,
+					initializationTimestamp,
+					0,
+					0,
+					jobStatus,
+					0,
+					tasksPerState,
+					0));
+			}
+		}
+	}
+
+	public CompletableFuture<Acknowledge> cancel(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.cancel(timeout));
+			} else {
+				log.info("Cancellation during initialization requested for job {}. Job will be cancelled once JobManager has been initialized.", jobId);
+				jobStatus = JobStatus.CANCELLING;
+				return jobResultFuture.thenApply(ignored -> Acknowledge.get());
+			}
+		}
+	}
+
+	public CompletableFuture<JobStatus> requestJobStatus(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobStatus(timeout));
+			} else {
+				return CompletableFuture.completedFuture(jobStatus);
+			}
+		}
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> requestJob(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJob(timeout));
+			} else {
+				return CompletableFuture.supplyAsync(() -> ArchivedExecutionGraph.createFromInitializingJob(jobId, jobName, null, jobStatus, initializationTimestamp));
+			}
```

Comment:
I think it will return the right result, but the naming of methods / variables here is terrible: `jobStatus == JobStatus.RUNNING` as soon as the `jobManagerRunnerFuture` completed. The job is not necessarily in RUNNING state. It could be in `CREATED` or `FAILING` ... we would still do the correct thing in `DispatcherJob`, but it is not very obvious.

I will improve the tracking of the job status inside the `DispatcherJob`, in particular the names (this seems to be the biggest flaw of my proposed changes).

Reply:
Well, you are right, the result will still be wrong. We need to delegate to the resultFuture when it's complete.